daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala

```diff
@@ -27,52 +28,69 @@ private[lf] object PartialTransaction {
   type Node = Node.GenNode[NodeId, Value.ContractId]
   type LeafNode = Node.LeafOnlyNode[Value.ContractId]
 
+  sealed abstract class ContextInfo {
+    private[PartialTransaction] def childrenSeeds(idx: Int): crypto.Hash
+  }
+
+  sealed abstract class RootContextInfo extends ContextInfo
+
+  private[PartialTransaction] final class SeededTransactionRootContext(seed: crypto.Hash)
+      extends RootContextInfo {
+    override private[PartialTransaction] def childrenSeeds(idx: Int) =
+      crypto.Hash.deriveNodeSeed(seed, idx)
+  }
+
+  private[PartialTransaction] final class SeededPartialTransactionRootContext(
+      seeds: ImmArray[Option[crypto.Hash]]
+  ) extends RootContextInfo {
+    override def childrenSeeds(idx: NodeIdx): Hash = {
+      val opt = if (0 <= idx && idx < seeds.length) {
+        seeds(idx)
+      } else {
+        None
+      }
+      opt.getOrElse(throw new RuntimeException(s"seed for ${idx}th root node not provided"))
+    }
+  }
+
+  private[PartialTransaction] object NoneSeededTransactionRootContext extends RootContextInfo {
+    override private[PartialTransaction] def childrenSeeds(idx: NodeIdx): Nothing = {
+      throw new RuntimeException(s"the machine is not configure to create transaction")
+    }
+  }
+
   /** Contexts of the transaction graph builder, which we use to record
     * the sub-transaction structure due to 'exercises' statements.
     */
   final class Context private (
-      val exeContext: Option[ExercisesContext], // empty if root context
-      val children: BackStack[NodeId],
-      protected val childrenSeeds: Int => crypto.Hash,
+      val info: ContextInfo,
+      val children: BackStack[NodeId] = BackStack.empty,
   ) {
     def addChild(child: NodeId): Context =
-      new Context(exeContext, children :+ child, childrenSeeds)
-    def nextChildrenSeed: crypto.Hash =
-      childrenSeeds(children.length)
+      new Context(info, children :+ child)
+    // needs to be lazy as info.childrenSeeds may be undefined on children.length
+    lazy val nextChildrenSeed: crypto.Hash =
```

Comment:
we only expect to call this once right? if so, I think the `def` might be less confusing while having the same performance.

And just for my own understanding. The case where this crashes is during validation when we have a partial transaction context and we find more nodes during reinterpretation than the original (partial) transaction contained?

Reply:
> we only expect to call this once right?

No. 

>  if so, I think the `def` might be less confusing while having the same performance.

I change the code so we used it at most one, and then change to `def`.

> And just for my own understanding. The case where this crashes is during validation when we have a partial transaction context and we find more nodes during reinterpretation than the original (partial) transaction contained?

Yes.